xlsynth-g8r/src/equiv/easy_smt_backend.rs

```diff
@@ -0,0 +1,600 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#![cfg(feature = "has-easy-smt")]
+
+use std::{
+    io,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+
+use easy_smt::{Context, ContextBuilder, SExpr};
+
+use crate::{
+    equiv::solver_interface::{BitVec, Solver},
+    ir_value_utils::{ir_bits_from_lsb_is_0, ir_value_from_bits_with_type},
+    test_solver,
+    xls_ir::ir,
+};
+
+#[derive(Debug, Clone)]
+pub struct SolverFn {
+    pub push_fn: fn(&mut Context) -> io::Result<()>,
+    pub pop_fn: fn(&mut Context) -> io::Result<()>,
+    pub check_fn: fn(&mut Context) -> io::Result<easy_smt::Response>,
+    pub assert_fn: fn(&mut Context, SExpr) -> io::Result<()>,
+}
+
+impl SolverFn {
+    pub fn default() -> Self {
+        Self {
+            push_fn: Context::push,
+            pop_fn: Context::pop,
+            check_fn: Context::check,
+            assert_fn: Context::assert,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct EasySMTConfig {
+    pub solver_path: PathBuf,
+    pub solver_args: Vec<String>,
+    pub replay_file: Option<PathBuf>,
+    pub solver_fn: SolverFn,
+}
+
+impl EasySMTConfig {
+    pub fn bitwuzla() -> Self {
+        Self {
+            solver_path: PathBuf::from("bitwuzla"),
+            solver_args: ["--produce-models".to_string()]
+                .iter()
+                .map(|s| s.to_string())
+                .collect(),
+            replay_file: None,
+            solver_fn: SolverFn {
+                push_fn: |ctx| ctx.push_many(1),
+                pop_fn: |ctx| ctx.pop_many(1),
+                ..SolverFn::default()
+            },
+        }
+    }
+
+    pub fn boolector() -> Self {
+        Self {
+            solver_path: PathBuf::from("boolector"),
+            solver_args: [
+                "--smt2",
+                "-m",
+                "--output-format=smt2",
+                "--no-exit-codes",
+                "--incremental",
+            ]
+            .iter()
+            .map(|s| s.to_string())
+            .collect(),
+            replay_file: None,
+            solver_fn: SolverFn {
+                push_fn: |ctx| ctx.push_many(1),
+                pop_fn: |ctx| ctx.pop_many(1),
+                ..SolverFn::default()
+            },
+        }
+    }
+
+    pub fn z3() -> Self {
+        Self {
+            solver_path: PathBuf::from("z3"),
+            solver_args: ["-nw", "-smt2", "-in"]
```

Comment:
Could you document the purpose for the flags that are not immediately intuitive (e.g. I'm not sure what -nw is)